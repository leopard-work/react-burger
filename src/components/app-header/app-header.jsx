import React from "react";
import {NavLink} from "react-router-dom";

import "@ya.praktikum/react-developer-burger-ui-components";
import { Logo, BurgerIcon, ListIcon, ProfileIcon } from "@ya.praktikum/react-developer-burger-ui-components";
import headerStyles from "./app-header.module.css";

const AppHeader = () => {
    return (
      <header className={headerStyles.header}>
          <div className="container">
              <div className={headerStyles.header_blocks + " pt-4 pb-4"}>
                  <NavLink to="/" className={headerStyles.logo}>
                      <Logo />
                  </NavLink>
                  <div className={headerStyles.left_btns}>
                      <NavLink to="/" exact className={isActive => headerStyles.btn + " p-4 mr-2 " + (isActive ? headerStyles.btn_active : '')}>
                          <BurgerIcon type="secondary" />
                          <span className="text text_type_main-default ml-2">Конструктор</span>
                      </NavLink>
                      <NavLink to="/orders" className={headerStyles.btn + " p-4 mr-2"}>
                          <ListIcon type="secondary" />
                          <span className="text text_type_main-default ml-2">Лента заказов</span>
                      </NavLink>
                  </div>
                  <NavLink to="/profile" className={isActive => headerStyles.btn + " p-4 mr-2 " + (isActive ? headerStyles.btn_active : '')}>
                      <ProfileIcon type="secondary" />
                      <span className="text text_type_main-default ml-2">Личный кабинет</span>
                  </NavLink>
              </div>
          </div>
      </header>
    );
}

export default AppHeader;
