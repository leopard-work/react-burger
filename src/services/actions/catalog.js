import { loadIngredients } from "../../utils/api";

export const GET_ITEMS_REQUEST = "GET_ITEMS_REQUEST";
export const GET_ITEMS_SUCCESS = "GET_ITEMS_SUCCESS";
export const GET_ITEMS_FAILED = "GET_ITEMS_FAILED";
export const CHANGE_ACTIVE_TAB = "CHANGE_ACTIVE_TAB";

export function getItems() {
  // Загрузка меню
  return function (dispatch) {
    dispatch({ type: GET_ITEMS_REQUEST });
    loadIngredients()
      .then((data) => {
        dispatch({ type: GET_ITEMS_SUCCESS, items: data });
      })
      .catch(() => {
        dispatch({ type: GET_ITEMS_FAILED });
      });
  };
}
