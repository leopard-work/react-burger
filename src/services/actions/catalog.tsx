import { loadIngredients } from "../../utils/api";
import { Dispatch } from "redux";

export const GET_ITEMS_REQUEST: "GET_ITEMS_REQUEST" = "GET_ITEMS_REQUEST";
export const GET_ITEMS_SUCCESS: "GET_ITEMS_SUCCESS" = "GET_ITEMS_SUCCESS";
export const GET_ITEMS_FAILED: "GET_ITEMS_FAILED" = "GET_ITEMS_FAILED";
export const CHANGE_ACTIVE_TAB: "CHANGE_ACTIVE_TAB" = "CHANGE_ACTIVE_TAB";

interface GetItemsRequestAction {
  readonly type: typeof GET_ITEMS_REQUEST;
}
interface GetItemsFailedAction {
  readonly type: typeof GET_ITEMS_FAILED;
}
interface GetItemsSuccessAction {
  readonly type: typeof GET_ITEMS_SUCCESS;
}

type GetItemsActions =
  | GetItemsRequestAction
  | GetItemsFailedAction
  | GetItemsSuccessAction;

type GetItemsDispatch = Dispatch<GetItemsActions>;

export function getItems() {
  // Загрузка меню
  return function (dispatch: GetItemsDispatch) {
    dispatch({ type: GET_ITEMS_REQUEST });
    loadIngredients()
      .then((data) => {
        dispatch({ type: GET_ITEMS_SUCCESS, items: data });
      })
      .catch(() => {
        dispatch({ type: GET_ITEMS_FAILED });
      });
  };
}
