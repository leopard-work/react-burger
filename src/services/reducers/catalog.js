import {
  CHANGE_ACTIVE_TAB,
  GET_ITEMS_FAILED,
  GET_ITEMS_REQUEST,
  GET_ITEMS_SUCCESS,
} from "../actions/catalog";

const initialState = {
  items: {
    // Меню
    data: [],
    success: false,
  },
  itemsRequest: true,
  itemsFailed: false,

  activeTab: "one", // Активаная вкладка навигации в меню
};

export const catalogReducer = (state = initialState, action) => {
  switch (action.type) {
    case GET_ITEMS_REQUEST: {
      return {
        ...state,
        itemsRequest: true,
      };
    }
    case GET_ITEMS_SUCCESS: {
      return {
        ...state,
        items: action.items,
        itemsRequest: false,
        itemsFailed: false,
      };
    }
    case GET_ITEMS_FAILED: {
      return {
        ...state,
        itemsFailed: true,
        itemsRequest: false,
      };
    }
    case CHANGE_ACTIVE_TAB: {
      // Установка активного элемента в навигации меню
      return {
        ...state,
        activeTab: action.activeTab,
      };
    }
    default: {
      return state;
    }
  }
};
