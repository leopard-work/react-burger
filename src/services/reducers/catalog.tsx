import {
  CHANGE_ACTIVE_TAB,
  GET_ITEMS_ACTIONS,
  GET_ITEMS_FAILED,
  GET_ITEMS_REQUEST,
  GET_ITEMS_SUCCESS,
} from "../actions/catalog";
import { ItemProps } from "../../utils/types";

type catalogState = {
  items: {
    data: Array<ItemProps>;
    success: boolean;
  };
  itemsRequest: boolean;
  itemsFailed: boolean;
  activeTab: string;
};

export const initialState: catalogState = {
  items: {
    // Меню
    data: [],
    success: false,
  },
  itemsRequest: true,
  itemsFailed: false,

  activeTab: "one", // Активаная вкладка навигации в меню
};

export const catalogReducer = (
  state = initialState,
  action: GET_ITEMS_ACTIONS
) => {
  switch (action.type) {
    case GET_ITEMS_REQUEST: {
      return {
        ...state,
        itemsRequest: true,
      };
    }
    case GET_ITEMS_SUCCESS: {
      return {
        ...state,
        items: action.items,
        itemsRequest: false,
        itemsFailed: false,
      };
    }
    case GET_ITEMS_FAILED: {
      return {
        ...state,
        itemsFailed: true,
        itemsRequest: false,
      };
    }
    case CHANGE_ACTIVE_TAB: {
      // Установка активного элемента в навигации меню
      return {
        ...state,
        activeTab: action.activeTab,
      };
    }
    default: {
      return state;
    }
  }
};
