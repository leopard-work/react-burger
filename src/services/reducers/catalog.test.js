import {
  CHANGE_ACTIVE_TAB,
  GET_ITEMS_FAILED,
  GET_ITEMS_REQUEST,
  GET_ITEMS_SUCCESS,
} from "../actions/catalog";

const testItem = {
  _id: "",
  name: "testItem",
  type: "",
  proteins: 0,
  fat: 0,
  carbohydrates: 0,
  calories: 0,
  price: 0,
  image: "",
  image_mobile: "",
  image_large: "",
  __v: 0,
  count: 0,
};
const testItems = [testItem, testItem, testItem];

import { initialState, catalogReducer } from "./catalog";
import renderer from "react-test-renderer";
import { Link } from "../../components/app-header/app-header";
import React from "react";

describe("Каталог товаров", () => {
  it("- исходное состояние", () => {
    expect(catalogReducer(undefined, {})).toEqual(initialState);
    const tree = renderer.create(<>Исходное состояние</>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("- обработка GET_ITEMS_REQUEST", () => {
    expect(
      catalogReducer(undefined, {
        type: GET_ITEMS_REQUEST,
      })
    ).toEqual({
      ...initialState,
      itemsRequest: true,
    });
    const tree = renderer.create(<>Запрос каталога</>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("- обработка GET_ITEMS_SUCCESS", () => {
    expect(
      catalogReducer(undefined, {
        type: GET_ITEMS_SUCCESS,
        items: testItems,
      })
    ).toEqual({
      ...initialState,
      items: testItems,
      itemsRequest: false,
      itemsFailed: false,
    });
    const tree = renderer.create(<>Получение товаров</>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("- обработка GET_ITEMS_FAILED", () => {
    expect(
      catalogReducer(undefined, {
        type: GET_ITEMS_FAILED,
      })
    ).toEqual({
      ...initialState,
      itemsFailed: true,
      itemsRequest: false,
    });
    const tree = renderer.create(<>Ошибка загрузки товаров</>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("- обработка CHANGE_ACTIVE_TAB", () => {
    expect(
      catalogReducer(undefined, {
        type: CHANGE_ACTIVE_TAB,
        activeTab: "sauce",
      })
    ).toEqual({
      ...initialState,
      activeTab: "sauce",
    });
    const tree = renderer
      .create(<>Замена вкладки категории в каталоге</>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
