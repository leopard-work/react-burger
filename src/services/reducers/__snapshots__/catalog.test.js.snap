// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Каталог товаров - исходное состояние 1`] = `"Исходное состояние"`;

exports[`Каталог товаров - обработка CHANGE_ACTIVE_TAB 1`] = `"Замена вкладки категории в каталоге"`;

exports[`Каталог товаров - обработка GET_ITEMS_FAILED 1`] = `"Ошибка загрузки товаров"`;

exports[`Каталог товаров - обработка GET_ITEMS_REQUEST 1`] = `"Запрос каталога"`;

exports[`Каталог товаров - обработка GET_ITEMS_SUCCESS 1`] = `"Получение товаров"`;
